dll/win32/shell32/shlexec.cpp

```diff
@@ -777,56 +777,29 @@ static UINT SHELL_FindExecutable(LPCWSTR lpPath, LPCWSTR lpFile, LPCWSTR lpVerb,
     }
 
     GetCurrentDirectoryW(ARRAY_SIZE(curdir), curdir);
-    if (!PathIsFileSpecW(lpFile))
+    if (lpPath && *lpPath)
     {
-        BOOL found = FALSE;
-        if (lpPath && *lpPath)
-        {
-            TRACE("lpPath %s\n", debugstr_w(lpPath));
-            PathCombineW(xlpFile, lpPath, lpFile);
-            if (PathFileExistsDefExtW(xlpFile, WHICH_DEFAULT | WHICH_OPTIONAL) || PathFileExistsW(xlpFile))
-            {
-                GetFullPathNameW(xlpFile, ARRAY_SIZE(xlpFile), xlpFile, NULL);
-                found = TRUE;
-            }
-        }
-        if (!found)
-        {
-            lstrcpyW(xlpFile, lpFile);
-            if (PathFileExistsDefExtW(xlpFile, WHICH_DEFAULT | WHICH_OPTIONAL) || PathFileExistsW(xlpFile))
-            {
-                GetFullPathNameW(xlpFile, ARRAY_SIZE(xlpFile), xlpFile, NULL);
-                found = TRUE;
-            }
-        }
-        if (found)
-        {
-            lpFile = xlpFile;
-            lstrcpyW(lpResult, xlpFile);
-        }
-        else
-            xlpFile[0] = '\0';
+        search_paths[0] = lpPath;
+        search_paths[1] = curdir;
     }
     else
+        search_paths[0] = curdir;
+    lstrcpyW(xlpFile, lpFile);
+    if (PathResolveW(xlpFile, search_paths, PRF_TRYPROGRAMEXTENSIONS | PRF_VERIFYEXISTS) ||
+        PathFindOnPathW(xlpFile, search_paths))
     {
-        if (lpPath && *lpPath)
-        {
-            search_paths[0] = lpPath;
-            search_paths[1] = curdir;
-        }
-        else
-            search_paths[0] = curdir;
-        lstrcpyW(xlpFile, lpFile);
-        if (PathResolveW(xlpFile, search_paths, PRF_TRYPROGRAMEXTENSIONS | PRF_VERIFYEXISTS))
-        {
-            TRACE("PathResolveW returned non-zero\n");
-            lpFile = xlpFile;
-            lstrcpyW(lpResult, xlpFile);
-            /* The file was found in lpPath or one of the directories in the system-wide search path */
-        }
-        else
-            xlpFile[0] = '\0';
+        TRACE("PathResolveW returned non-zero\n");
+        lpFile = xlpFile;
+        lstrcpyW(lpResult, xlpFile);
+        /* The file was found in lpPath or one of the directories in the system-wide search path */
     }
+    else
+    {
+        xlpFile[0] = '\0';
+    }
+
+    /* Clear any leading/trailing spaces */
+    PathRemoveBlanksW(xlpFile);
```

Comment:
This could be moved inside the if-block above (because in the else case you just NUL the string so there wouldn't be any leading/trailing spaces to remove anymore).
Also, should this clearing be done before invoking the `lstrcpyW(lpResult, xlpFile);` ?

Reply:
>This could be moved inside the if-block above (because in the else case you just NUL the string so there wouldn't be any leading/trailing spaces to remove anymore).

Indeed, this is even more correct.

>Also, should this clearing be done before invoking the lstrcpyW(lpResult, xlpFile); ?

Yeah, it should, because as I understand, that first copy to output `lpResult` stores an unmodified path for the case anything else from below fails (so it remains unmodified later).

Reply:
Applied.